### PR TITLE
[feat] Add Memory Layout Module

### PR DESCRIPTION
**Pull request by ajisaxena**
This PR resolved the following:-

1. Makes the linker script simpler by removing the variables needed by the code. 
2. One place to have all the memory related variables. 
3. Added test cases such that invalid variables are caught as part of unit testing. 
4. At many occasions it was observed the data structure was bigger than the size of memory reserved for it. Since all the memory related variables are in single module, it is easier to catch these type of issues in unit test. 

Please note that with this PR I have moved the ROM to use the memory_layout module. I would also recommend other components [FMC and RT] to start using this.

### CODE REVIEW

**Review by rusty1968 — DISMISSED** [review]

**Review by mhatrevi — CHANGES_REQUESTED** [review]

File: rom/dev/src/fuse.rs [hunk 1]
```diff
@@ -48,8 +44,10 @@ pub fn log_fuse_data(entry_id: FuseLogEntryId, data: &[u8]) -> CaliptraResult<()
     };
     log_entry.log_data.as_bytes_mut()[..data.len()].copy_from_slice(data);
 
-    // Get a pointer to the FUSE log. This is a static array of FUSE log entries.
-    let dst = unsafe { &mut FUSE_LOG_ORG[..] };
+    let dst = unsafe {
+        let ptr = FUSE_LOG_ORG as *mut FuseLogEntry;
+        core::slice::from_raw_parts_mut(ptr, FUSE_LOG_SIZE)
```
[the remark this answers is not in the document]
Comment (ajisaxena) [reply rom/dev/src/fuse.rs:49]: Yea, I was planning to have the size as number of entries but missed it.

File: rom/dev/src/flow/cold_reset/mod.rs [hunk 2]
```diff
@@ -99,23 +95,23 @@ pub fn copy_tbs(tbs: &[u8], tbs_type: TbsType, env: &mut RomEnv) -> CaliptraResu
...
-                let ptr = &mut LDEVID_TBS_ORG as *mut u8;
+                let ptr = LDEVID_TBS_ORG as *mut u8;
                 core::slice::from_raw_parts_mut(ptr, tbs.len())
             }
         }
         TbsType::FmcaliasTbs => {
             env.fht_data_store.fmcalias_tbs_size = tbs.len() as u16;
             unsafe {
-                let tbs_max_size = &FMCALIAS_TBS_SIZE as *const u8 as usize;
+                let tbs_max_size = FMCALIAS_TBS_SIZE as *const u8 as usize;
```
[the remark this answers is not in the document]
Comment (ajisaxena) [reply rom/dev/src/flow/cold_reset/mod.rs]: Updated.

File: rom/dev/src/fuse.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajisaxena) [reply rom/dev/src/fuse.rs]: Updated.

File: rom/dev/src/pcr.rs [hunk 3]
```diff
@@ -128,8 +125,10 @@ pub fn log_pcr(pcr_entry_id: PcrLogEntryId, pcr_id: PcrId, data: &[u8]) -> Calip
     };
     pcr_log_entry.pcr_data.as_bytes_mut()[..data.len()].copy_from_slice(data);
 
-    // Get a pointer to the PCR log. This is a static array of PCR log entries.
-    let dst = unsafe { &mut PCR_LOG_ORG[..] };
+    let dst = unsafe {
+        let ptr = PCR_LOG_ORG as *mut PcrLogEntry;
+        core::slice::from_raw_parts_mut(ptr, PCR_LOG_SIZE)
```
[the remark this answers is not in the document]
Comment (ajisaxena) [reply rom/dev/src/pcr.rs]: Updated

File: common/src/memory_layout.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+#[allow(clippy::assertions_on_constants)]
+fn mem_layout_test_manifest() {
+    assert!(MAN1_SIZE as usize >= core::mem::size_of::<ImageManifest>());
+    assert!(MAN2_SIZE as usize >= core::mem::size_of::<ImageManifest>());
+    if MAN1_SIZE != MAN2_SIZE {
+        assert!(false);
+    }
+
+    assert!((MAN2_ORG - MAN1_ORG) == MAN1_SIZE);
+    assert!((MAN2_ORG - MAN1_ORG) as usize >= core::mem::size_of::<ImageManifest>());
```
[the remark this answers is not in the document]
Comment (ajisaxena) [reply common/src/memory_layout.rs]: Updated

File: common/src/memory_layout.rs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+pub const STACK_SIZE: u32 = 14 * 1024;
+pub const ESTACK_SIZE: u32 = 1024;
+pub const NSTACK_SIZE: u32 = 1024;
+
+#[test]
+#[allow(clippy::assertions_on_constants)]
+fn mem_layout_test_manifest() {
+    assert!(MAN1_SIZE as usize >= core::mem::size_of::<ImageManifest>());
+    assert!(MAN2_SIZE as usize >= core::mem::size_of::<ImageManifest>());
+    if MAN1_SIZE != MAN2_SIZE {
```
[the remark this answers is not in the document]
Comment (ajisaxena) [reply common/src/memory_layout.rs]: Updated

**Review by mhatrevi — APPROVED** [review]

**Review by rusty1968 — APPROVED** [review]

**Comment by rusty1968** [comment]
Can this be merged now?

**Review by jhand2 — APPROVED** [review]

**Pull request merged**